perception/lidar_transfusion/lib/transfusion_trt.cpp

```diff
@@ -0,0 +1,262 @@
+      static_cast<int32_t>(params_input_cpu), static_cast<int32_t>(config_.num_point_values_)});
+  network_trt_ptr_->context->setTensorAddress(
+    network_trt_ptr_->getTensorName(NetworkIO::cls_score), cls_output_d_.get());
+  network_trt_ptr_->context->setTensorAddress(
+    network_trt_ptr_->getTensorName(NetworkIO::bbox_pred), box_output_d_.get());
+  network_trt_ptr_->context->setTensorAddress(
+    network_trt_ptr_->getTensorName(NetworkIO::dir_pred), dir_cls_output_d_.get());
+  return true;
+}
+
+bool TransfusionTRT::inference()
+{
+  auto status = network_trt_ptr_->context->enqueueV3(stream_);
+  CHECK_CUDA_ERROR(cudaStreamSynchronize(stream_));
+
+  if (!status) {
+    RCLCPP_WARN_STREAM(
+      rclcpp::get_logger("lidar_transfusion"), "Fail to enqueue and skip to detect.");
+    return false;
+  }
+  return true;
+}
+
+bool TransfusionTRT::postprocess(std::vector<Box3D> & det_boxes3d)
+{
+  CHECK_CUDA_ERROR(post_ptr_->generateDetectedBoxes3D_launch(
+    cls_output_d_.get(), box_output_d_.get(), dir_cls_output_d_.get(), det_boxes3d, stream_));
+  CHECK_CUDA_ERROR(cudaStreamSynchronize(stream_));
+  return true;
+}
+
+bool TransfusionTRT::postprocessCPU(std::vector<Box3D> & det_boxes3d)
+{
+  std::vector<float> cls_output_cpu(cls_size_);
+  std::vector<float> box_output_cpu(box_size_);
+  std::vector<float> dir_cls_output_cpu(dir_cls_size_);
+
+  CHECK_CUDA_ERROR(cudaMemcpyAsync(
+    cls_output_cpu.data(), cls_output_d_.get(), cls_size_ * sizeof(float), cudaMemcpyDeviceToHost,
+    stream_));
+  CHECK_CUDA_ERROR(cudaMemcpyAsync(
+    box_output_cpu.data(), box_output_d_.get(), box_size_ * sizeof(float), cudaMemcpyDeviceToHost,
+    stream_));
+  CHECK_CUDA_ERROR(cudaMemcpyAsync(
+    dir_cls_output_cpu.data(), dir_cls_output_d_.get(), dir_cls_size_ * sizeof(float),
+    cudaMemcpyDeviceToHost, stream_));
+  CHECK_CUDA_ERROR(cudaStreamSynchronize(stream_));
+
+  for (size_t i = 0; i < config_.num_proposals_; ++i) {
+    auto class_id = 0;
+    auto max_score = cls_output_cpu[i];
+    for (size_t j = 1; j < config_.num_classes_; ++j) {
+      auto score = cls_output_cpu[config_.num_proposals_ * j + i];
+      if (max_score < score) {
+        max_score = score;
+        class_id = j;
+      }
+    }
+    if (max_score < config_.score_threshold_) {
+      continue;
+    }
+    det_boxes3d.emplace_back(lidar_transfusion::Box3D{
+      class_id, max_score,
+      box_output_cpu[i] * config_.num_point_values_ * config_.voxel_x_size_ + config_.min_x_range_,
+      box_output_cpu[i + config_.num_proposals_] * config_.num_point_values_ *
+          config_.voxel_y_size_ +
+        config_.min_y_range_,
+      box_output_cpu[i + 2 * config_.num_proposals_],
+      std::exp(box_output_cpu[i + 3 * config_.num_proposals_]),
+      std::exp(box_output_cpu[i + 4 * config_.num_proposals_]),
+      std::exp(box_output_cpu[i + 5 * config_.num_proposals_]),
+      std::atan2(dir_cls_output_cpu[i], dir_cls_output_cpu[i + config_.num_proposals_])});
```

Comment:
In centerpoint, we discovered that the values from the cos/sin estimation do not always have a unit norm (why would they anyways, ML does not assure that).
There are cases where the objects have orientation error of 90 degrees, that causes several overrides. We found out that these cases are highly correlated to the cases where cos/sin values are both close to 0 (or have a norm quite different to 1). I implemented a simple filter for that ages ago for centerpoint. Can you evaluate if that phenomena happens in transfusion? If so, can you reimplement that functionality in transfusion?

(I commented this in the cpu implementation rather than the cuda one due to the review order, but it also applies there as well)

Reply:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/7a054cca63abac2ac20990cfff676c22ff92e6cc
Right, I forgot to check it for host processing. Moreover, I checked the output with  `yaw_norm < config_.yaw_norm_thresholds_[class_id]` condition and some frames generate detection where `max_score >= config_.score_threshold_`.  These detection had 90 degrees orientation error. Therefore we need that filter for `lidar_transfusion` as well.